Added Approximated dbscan to linfa-clustering

Sauro98 opened the pull request:
I added an implementation of the Approximated DBSCAN to the linfa-clustering subcrate. This algorithm depends on the crate `partitions` for an implementation of the union-find structure. This can be removed by using adjacency lists, at the cost of performance. Another big thing: I noticed that the existing vanilla DBSCAN implementation used an axis convention that was the opposite to the one used by the `generate_blobs` function, which it uses in the benches. I modified the DBSCAN implementation to invert the axes in order to be able to compare benches of the two implementations, and I also modified the tests to pass this modification. I modified the vanilla implementation only to be able to compare the performances, I did not want to judge the previous implementation. This was also my first time using `ndarray` so any suggestion is very welcome.

bytesnake commented:
Thank you for the PR! I will do a fast look through now, but for a deeper understanding I will need more time. End of next week should be possible for a merge :)

Sauro98 commented on linfa-clustering/src/dbscan/algorithm.rs:
```diff
@@ -147,19 +147,20 @@ mod tests {
         let params = DbscanHyperParams::new(2).tolerance(1.0).build();
 
-        let mut data: Array2<f64> = Array2::zeros((2, 50));
+        let mut data: Array2<f64> = Array2::zeros((50, 2));
```
The DBSCAN implementation accepted input shaped as (features, number of points), while `generate_blobs` generates data shaped as (number of points, features) so I modified the DBSCAN implementation to match `generate_blobs` in order to be able to bench and so I also had to change the tests

Sauro98 commented on linfa-clustering/src/appx_dbscan/cell/mod.rs:
```diff
@@ -0,0 +1,231 @@
+    cell.points.iter().filter(|s_p| points.row(point_i).l2_dist(&points.row(s_p.point_index)).unwrap() <= tolerance).count()
+}
+
+#[cfg(test)]
```
The `mod tests;` line appears in the following line and references the file `tests.rs` in the same folder. Running `cargo tests` does seem to run the tests included in that file so I assume it is working. I don't know if this is what you meant.

bytesnake replied:
that was just uncommon to my eye, but totally fine

bytesnake commented:
Could you take a look at https://github.com/rust-ml/linfa/blob/fdf241127fe9058ae98c67a603a2bbb8eccc561c/CONTRIBUTE.md, I'm currently writing a small how-to-contribute document for the upcoming interface changes. If you find something unclear, please let me know so I can update the document. The interface is a bit different, to the original K-Means implementation because of the new traits. If you find some drawbacks in ergonomic, please let me know as well.

Sauro98 commented:
I gave it a quick read and it looks clear to me. Are the suggested changes already implementable? If so in the next days I will try to apply them to the `appx_dbscan` algorithm and commit them to this PR to see if I got them right. If so I could apply them to the other clustering algorithms too if needed.

bytesnake commented:
yes that would be awesome! I'm still waiting for feedback of @paulkoerbitz but I have moved the PR to a separate branch `traits`. I would suggest that in order to move forward you rebase your PR to this branch and we will merge everything in one go later on.

Sauro98 commented:
Hi, I've changed the interface according to the traits, added some documentation, and some improvements. I rebased the PR on the traits branch to do this, should I go back to master now?

bytesnake commented:
yes the traits branch was merged into master in the meantime

Sauro98 commented:
Man, this git stuff is hard. Now everything works fine (at least on my side) but there are 240 commits here now. I don't know if this is fine or how/if to clean this up. I (think) I'm done working on the `appx_dbscan`. I'm here for any comments about the changes I brought in and to help clean this up if possible. I only changed the contents of `linfa_clustering` so if there's any other file outside that's modified then it happened when I messed around with rebasing.

bytesnake commented:
:see_no_evil: :see_no_evil: have you tried a `git rebase --onto master`? Otherwise write me an email [here](mailto:bytesnake@mailbox.org) and we will figure something out

bytesnake commented:
I would recommend to follow this document https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request Especially squashing your commits into one single makes the rebase much more simple. Also you have to add the upstream remote before chosing the right commit.

Sauro98 commented:
Hi, so I somehow managed to get it all down to just one commit based on master (well two counting the formatting one). I don't understand why code coverage is at 0% when there are definitely some tests in place. I want to thank you a lot for the guide you provided me, I was lost in the sea of articles about rebasing but with that one, I finally understand (I think) how it works.

bytesnake reviewed (commented):
added some comments on the code style and some edges which stand out to me. Otherwise the algorithm is a bit too much to understand in detail for me right now, but if you show equivalence with DBSCAN then I'm fine with merging. (but have at least three different tests for the algorithm)

bytesnake commented:
no worries about the codecov report, this is at the moment only enabled for the base crate because of some issues we have https://github.com/rust-ml/linfa/issues/35

Sauro98 commented on linfa-clustering/Cargo.toml:
```diff
@@ -23,6 +23,11 @@ version = "1.0"
 default-features = false
 features = ["std", "derive"]
 
+[features]
```
Oh ok, I couldn't get the tests to run due to lapacke errors and so I followed what was written on the contribution guide. I guess that means that there's some problem with the OpenBLAS installation on my machine

bytesnake replied:
what error do you get? Have you by any chance a archlinux system, because there is a problem with the blas libraries

Sauro98 replied:
I'm running a linux mint virtual machine. I get errors that some lapacke symbols are not found. I tried installing manually OpenBLAS following a guide and the installation went fine, but that still didn't solve my problems

bytesnake replied:
can you try adding the lapacke library explicitly: `RUSTFLAGS="-C link-arg=-llapacke -C link-arg=-lcblas" cargo test`

Sauro98 replied:
Ok, there was a problem on my installation that prevented the tests to run even when adding the library explicitly. Now I've fixed that and it works if I explicitly add the libraries but if I don't I get the same error as before.

bytesnake replied:
thought this was a special case for archlinux this definitely needs some more debugging. As a temporary fix you can add the RUSTFLAGS to your `.config` https://doc.rust-lang.org/cargo/reference/config.html

Sauro98 commented on linfa-clustering/benches/dbscan.rs:
```diff
@@ -18,11 +19,18 @@ fn dbscan_bench(c: &mut Criterion) {
+                black_box(
+                    Dbscan::params(min_points)
+                        .tolerance(tolerance)
+                        .build()
```
I'm sorry I don't understand what you mean by that

bytesnake replied:
the `build` call just checks the validity of hyper-parameters, you can also move that into the `transform` call and simplify the API a bit

Sauro98 replied:
Doesn't `Dbscan::params` return a builder and then `build()` returns the actual params for which `Transformer` is implemented? Or are you suggesting to implement `Transformer` for the builder where `build().transform(...)` is called?

bytesnake replied:
yes

Sauro98 replied:
Ok, I got it. Thank you for all your patience

Sauro98 commented on linfa-clustering/src/appx_dbscan/cell/mod.rs:
```diff
@@ -0,0 +1,296 @@
+        let index_arr = get_base_cell_index(&curr_point, params);
+        let cell = table
+            .entry(index_arr.clone())
+            .or_insert(Cell::new(&index_arr));
```
I think that file doesn't exist anymore. Now I keep integers in the table.

Sauro98 commented on linfa-clustering/src/appx_dbscan/tests.rs:
```diff
@@ -0,0 +1,112 @@
+use rand_isaac::Isaac64Rng;
+
+#[test]
+fn appx_dbscan_test_100() {
```
I'm sorry what do you mean by this? Do you mean checking that each noise point in the exact result is also a noise point in the approximated result?

bytesnake replied:
you could add a test with only one large cluster and than manually add some noise way outside the inner cluster. The DBSCAN algorithm should identify them as noise.

Sauro98 commented:
Hi, I think I made all the changes pointed out in the last review. There are 3 things that I'd like to point out:

* In order to implement `Transformer` which takes a reference `&self` I had to change the `build` method of the builder to take a reference as well
* In the equivalence tests between DBSCAN and the approximated version I had to use small datasets otherwise the unoptimized tests would take far too long
* The equivalence tests are done on datasets generated at random. Currently, the value `1e-4` for the slack is small enough to assure the convergence of the approximated result to the exact one, but there may be instances where particularly nasty datasets are created and the tests may need a smaller value of the slack to assure convergence.

Generally, the approximated DBSCAN algorithm with parameters `tolerance` and `slack` guarantees to return a result that is between the results of the exact DBSCAN with parameter `tolerance` as a lower bound and with parameter `tolerance * (1 + slack)` as an upper bound. That is, I should check that every approximated cluster is contained in the upper bound exact one and contains the lower bound exact one, but I don't know if it would be worth it to make such a check in the tests.

If you believe that such a check should be performed then I will try to implement it.

bytesnake commented:
> Hi, I think I made all the changes pointed out in the last review. There are 3 things that I'd like to point out:
> 
>     * In order to implement `Transformer` which takes a reference `&self` I had to change the `build` method of the builder to take a reference as well

:+1:
> 
>     * In the equivalence tests between DBSCAN and the approximated version I had to use small datasets otherwise the unoptimized tests would take far too long

can you modify the `.github/workflows/testing.yml` to build tests in release mode? I would prefer this rather than reduce testing data. 
> 
>     * The equivalence tests are done on datasets generated at random. Currently, the value `1e-4` for the slack is small enough to assure the convergence of the approximated result to the exact one, but there may be instances where particularly nasty datasets are created and the tests may need a smaller value of the slack to assure convergence.

your `Array2::random_using` is using a constant seed therefore the random data should be the same in every instance
> 
> 
> Generally, the approximated DBSCAN algorithm with parameters `tolerance` and `slack` guarantees to return a result that is between the results of the exact DBSCAN with parameter `tolerance` as a lower bound and with parameter `tolerance * (1 + slack)` as an upper bound. That is, I should check that every approximated cluster is contained in the upper bound exact one and contains the lower bound exact one, but I don't know if it would be worth it to make such a check in the tests.
> 
> If you believe that such a check should be performed then I will try to implement it.

mh I'm fine with the current number of tests, but I just opened an issue for ideas https://github.com/rust-ml/linfa/issues/65

Sauro98 commented:
I've added an outliers test and the check for nested clusters that is also done in the exact version. Now I've set the tests to run in release mode and so I've brought the test datasets to 1000 and 2000 points. Is that outliers test like the one you had in mind?

bytesnake reviewed (approved):
Thanks for all the work :tada:

bytesnake commented:
yep your outlier test was fine and implemented what I had in mind

Sauro98 commented:
Thank you, you've been very helpful! I'm sorry that I made you work so much for this PR haha
